### PR TITLE
Endpoints, exports and management for project views

### LINKED ISSUE #4064: Create endpoint that lists available views
Front-end needs an endpoint that returns a list of available views. Front-end needs some unique indentification (uid? url?) and a name to display

### LINKED ISSUE #4066: Add new permissions: view_metadata, change_metadata

### LINKED ISSUE #4067: Add new permission view_permissions
Questions:
- Should this be an assignable permission or implied by another (calculated only)?
  - If calculated, which permission implies this one? (assuming `manage_asset` for now)
    - #2956 has this change: "Consistently enforce the rule that users with explicitly-assigned permissions see only their own permissions and the owner's permissions, not other users' permissions" -- are we wanting to change this behavior?
- Management command to apply perm retroactively?

Checks:
- [ ] bulk perm assignments

### LINKED ISSUE #4068: Add a way of defining a custom assets view

### LINKED ISSUE #4079: Create endpoint for downloading projects list with metadata and users

### LINKED ISSUE #4081: An endpoint for getting assets for given view

### PR DESCRIPTION
**Pull request by joshuaberetta**
## Description

Allow for management of project views and assignments to users within the Django admin interface, as well as APIs to access filtered project views and asynchronously export the data to CSV — and be notified by email when export is complete. New permissions of `view_permissions` and `change_metadata` have also been added which can only be assigned at the project view level.

## Endpoints

The following new endpoints have been added:
- List user's available project views: `GET /api/v2/project-views` with example response of:
```json
        {
            "uid": "<uid>",
            "name": "Africa",
            "url": "<kf_url>/api/v2/project-views/<uid>/",
            "assets": "<kf_url>/api/v2/project-views/<uid>/assets/",
            "assets_count": 42,
            "assets_export": "<kf_url>/api/v2/project-views/<uid>/assets/export/",
            "users": "<kf_url>/api/v2/project-views/<uid>/users/",
            "users_export": "<kf_url>/api/v2/project-views/<uid>/users/export/",
            "countries": [
                "South Africa",
                "Zimbabwe",
                "Zambia",
                "Botswana"
            ],
            "permissions": [
                "view_asset"
            ],
            "assigned_users": [
                "bob",
                "alice"
            ]
        },
```
- Filtered list of assets for view, paginated: `GET /api/v2/project-views/<uid>/assets`
- Exports for filtered list of assets for view: 
  - Create new export: `POST /api/v2/project-views/<uid>/assets/export`
  - Get URL to export result (user also notified by email): `GET /api/v2/project-views/<uid>/assets/export`
- Filtered list of users for view, paginated: `GET /api/v2/project-views/<uid>/users`
- Exports for filtered list of users for view:
  - Create new export: `POST /api/v2/project-views/<uid>/users/export`
  - Get URL to export result (user also notified by email): `GET /api/v2/project-views/<uid>/users/export`

## Creating and assigning Project Views

A new project view can be created in the Django admin interface under "Project Views". The following fields are required:
- "Name": A valid string.
- "Countries": A list of comma-separated, three letter country codes.
- "Permissions"
  - `view_asset`:
    Minimum permissions required to list assets in project view. Shows a condensed asset list compared to `/api/v2/assets`
  - `view_submissions`:
    Allow users assigned to project view to access `/api/v2/assets/<uid>/data`
  - `view_permissions`:
    Allow users assigned to project view to see all user permissions assigned to each asset in the view.
  - `change_metadata`:
    Allow users assigned to project view to edit the metadata fields of an asset within the view. Only fields within `Asset.settings` and `Asset.name` are editable.

A view is then assigned on a per user basis under the "Assignments" section.

## Related issues

closes #4064 
closes #4066
closes #4067 
closes #4068 
closes #4079 
closes #4081

### CODE REVIEW

File: kpi/utils/project_views.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# coding: utf-8
+from __future__ import annotations
+
+from django.db.models import Q
+from rest_framework.request import Request
```
Comment (noliveleger) [inline kpi/utils/project_views.py:5]: `Request` is not used

File: kpi/serializers/v2/asset.py [hunk 2]
```diff
@@ -720,3 +745,86 @@ class AssetUrlListSerializer(AssetSerializer):
...
+        return obj.owner.email
+
+    def get_owner__name(self, obj: Asset) -> str:
+        return self._get_user_detail(obj, 'name')
+
+    def get_owner__organization(self, obj: Asset) -> str:
+        return self._get_user_detail(obj, 'organization')
+
+    def get_permissions(self, *args: list, **kwargs: dict) -> list:
+        if view_has_perm(self._get_view(), 'view_permissions'):
```
Comment (noliveleger) [inline kpi/serializers/v2/asset.py:817]: Should the user see the permissions of the asset if they are the owner even if the view does not that permission?

File: kpi/serializers/v2/asset.py [hunk 3]
```diff
@@ -720,3 +745,86 @@ class AssetUrlListSerializer(AssetSerializer):
 
     class Meta(AssetSerializer.Meta):
         fields = ('url',)
+
+
+class AssetMetadataListSerializer(AssetSerializer):
```
Comment (noliveleger) [inline kpi/serializers/v2/asset.py:750]: Should not it extend `AssetListSerializer` instead?

File: kobo/apps/project_views/views.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    def _get_regional_response(self, queryset, _type):
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self._get_regional_serializer(page, _type=_type)
+            return self.get_paginated_response(serializer.data)
+        serializer = self._get_regional_serializer(queryset, _type=_type)
+        return Response(serializer.data)
+
+    def _get_regional_serializer(self, queryset, _type):
+        serializer = AssetMetadataListSerializer
```
Comment (noliveleger) [inline kobo/apps/project_views/views.py:121]: The serializer context should contain `object_permissions_per_asset` property when serializer is `AssetMetadataListSerializer`. It optimizes the queries made to DB to retrieve permissions in assets list. 
 
See https://github.com/kobotoolbox/kpi/blob/eab99a8f6c695f4f3ab36b8e55d7394c3042925b/kpi/views/v2/asset.py#L594-L614

Obviously `AssetMetadataListSerializer` should extend  `AssetListSerializer`

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kpi/serializers/v2/asset.py [hunk 5]
```diff
@@ -168,8 +174,22 @@ class Meta:
...
     def update(self, asset, validated_data):
+        request = self.context['request']
+        user = request.user
+        if not asset.has_perm(
+            user, 'change_asset'
+        ) and user_has_regional_asset_perm(asset, user, 'change_metadata'):
+            _validated_data = {}
+            settings = validated_data.get('settings')
+            if settings:
+                _validated_data['settings'] = settings
```
[the remark this answers is not in the document]
Comment (joshuaberetta) [reply kpi/serializers/v2/asset.py:185]: Love a good opportunity for 🦭

File: kpi/serializers/v2/asset.py (hunk 2 above)
Comment (joshuaberetta) [reply kpi/serializers/v2/asset.py:817]: Hmm good point — probably!

File: kpi/serializers/v2/asset.py [hunk 6]
```diff
@@ -168,8 +174,22 @@ class Meta:
         }
 
     def update(self, asset, validated_data):
+        request = self.context['request']
+        user = request.user
+        if not asset.has_perm(
+            user, 'change_asset'
+        ) and user_has_regional_asset_perm(asset, user, 'change_metadata'):
```
[the remark this answers is not in the document]
Comment (joshuaberetta) [reply kpi/serializers/v2/asset.py:182]: Much nicer 🙃 you should fork `black` 🍴

File: kpi/permissions.py [hunk 7]
```diff
@@ -241,6 +248,14 @@ class IsOwnerOrReadOnly(permissions.DjangoObjectPermissions):
     perms_map['OPTIONS'] = perms_map['GET']
     perms_map['HEAD'] = perms_map['GET']
 
+    def has_object_permission(self, request, view, obj):
+        user = get_database_user(request.user)
+        if user_has_regional_asset_perm(
+            obj, user, 'change_metadata'
+        ):
```
[the remark this answers is not in the document]
Comment (joshuaberetta) [reply kpi/permissions.py:255]: Haha I feel like there was a good reason but will need to check again...

File: kpi/serializers/v2/asset.py [hunk 8]
```diff
@@ -720,3 +745,86 @@ class AssetUrlListSerializer(AssetSerializer):
...
+
+    def get_data(self, *args: list, **kwargs: dict) -> str:
+        if view_has_perm(self._get_view(), PERM_VIEW_SUBMISSIONS):
+            return super().get_data(*args, **kwargs)
+        return ''
+
+    def get_date_first_deployement(self, obj: Asset) -> Optional[datetime.datetime]:
+        version = obj.asset_versions.filter(deployed=True).last()
+        if version:
+            return version.date_modified
```
[the remark this answers is not in the document]
Comment (joshuaberetta) [reply kpi/serializers/v2/asset.py:800]: `black` didn't like the walrus operation in there for some reason, so just move it around:
```python
first_version = (
    obj.asset_versions.only('date_modified')
    .filter(deployed=True)
    .order_by('date_modified')
    .first()
)
if first_version:
    return first_version.date_modified
```

File: kobo/apps/project_views/views.py (hunk 4 above)
Comment (noliveleger) [reply kobo/apps/project_views/views.py:121]: kpi#4195

**Pull request merged**